core/lib/dal/src/data_availability_dal.rs

```diff
@@ -0,0 +1,201 @@
+use zksync_db_connection::{connection::Connection, error::DalResult, instrument::InstrumentExt};
+use zksync_types::{pubdata_da::DataAvailabilityBlob, L1BatchNumber};
+
+use crate::{
+    models::storage_data_availability::{L1BatchDA, StorageDABlob},
+    Core,
+};
+
+#[derive(Debug)]
+pub struct DataAvailabilityDal<'a, 'c> {
+    pub(crate) storage: &'a mut Connection<'c, Core>,
+}
+
+impl DataAvailabilityDal<'_, '_> {
+    /// Inserts the blob_id for the given L1 batch. If the blob_id is already present,
+    /// verifies that it matches the one provided in the function arguments
+    /// (preventing the same L1 batch from being stored twice)
+    pub async fn insert_l1_batch_da(
+        &mut self,
+        number: L1BatchNumber,
+        blob_id: &str,
+        sent_at: chrono::NaiveDateTime,
+    ) -> anyhow::Result<()> {
+        let update_result = sqlx::query!(
+            r#"
+            INSERT INTO
+                data_availability (l1_batch_number, blob_id, sent_at, created_at, updated_at)
+            VALUES
+                ($1, $2, $3, NOW(), NOW())
+            "#,
+            i64::from(number.0),
+            blob_id,
+            sent_at,
+        )
+        .instrument("insert_l1_batch_da")
+        .with_arg("number", &number)
+        .with_arg("blob_id", &blob_id)
+        .report_latency()
+        .execute(self.storage)
+        .await?;
+
+        if update_result.rows_affected() == 0 {
+            tracing::debug!(
+                "L1 batch #{number}: DA blob_id wasn't updated as it's already present"
+            );
+
+            // Batch was already processed. Verify that existing DA blob_id matches
+            let matched: i64 = sqlx::query!(
+                r#"
+                SELECT
+                    COUNT(*) AS "count!"
+                FROM
+                    data_availability
+                WHERE
+                    l1_batch_number = $1
+                    AND blob_id = $2
+                "#,
+                i64::from(number.0),
+                blob_id,
+            )
+            .instrument("get_matching_batch_da_blob_id")
+            .with_arg("number", &number)
+            .report_latency()
+            .fetch_one(self.storage)
+            .await?
+            .count;
+
+            anyhow::ensure!(
+                matched == 1,
+                "DA blob_id verification failed. DA blob_id for L1 batch #{number} does not match the expected value"
```

Comment:
I added the `blob_id` that is being queried here. I am not sure about the second one though, because logically if the returned number is not 1 - it's almost always going to be 0, meaning that the second blob_id is not present